File: pm_core/src/infrastructure/store/adapter/issue_type_adapter.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package adapter

import (
	"context"

	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/core/port/store"
	"github.com/serp/pm-core/src/infrastructure/store/mapper"
	"github.com/serp/pm-core/src/infrastructure/store/model"
	"gorm.io/gorm"
)

type IssueTypeAdapter struct {
	db     *gorm.DB
	mapper *mapper.IssueTypeMapper
}

func NewIssueTypeAdapter(db *gorm.DB) store.IIssueTypeStore {
	return &IssueTypeAdapter{
		db:     db,
		mapper: mapper.NewIssueTypeMapper(),
	}
}

func (i *IssueTypeAdapter) CreateIssueType(ctx context.Context, tx *gorm.DB, issueType *entity.IssueTypeEntity) (*entity.IssueTypeEntity, error) {
	db := i.getDB(tx)
	issueTypeModel := i.mapper.ToModel(issueType)
	if err := db.WithContext(ctx).Create(issueTypeModel).Error; err != nil {
		return nil, err
	}
	return i.mapper.ToEntity(issueTypeModel), nil
}

func (i *IssueTypeAdapter) ExistsTypeKey(ctx context.Context, tenantID int64, typeKey string) (bool, error) {
	var count int64
	if err := i.db.WithContext(ctx).Model(&model.IssueTypeModel{}).
		Where("tenant_id = ? AND type_key = ? AND deleted_at IS NULL", tenantID, typeKey).
		Count(&count).Error; err != nil {
		return false, err
	}
	return count > 0, nil
}

func (i *IssueTypeAdapter) GetIssueTypeByID(ctx context.Context, issueTypeID int64, tenantID int64) (*entity.IssueTypeEntity, error) {
	var issueTypeModel model.IssueTypeModel
	if err := i.db.WithContext(ctx).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", issueTypeID, tenantID).
		First(&issueTypeModel).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, err
	}
	return i.mapper.ToEntity(&issueTypeModel), nil
}

func (i *IssueTypeAdapter) ListIssueTypes(ctx context.Context, tenantID int64) ([]*entity.IssueTypeEntity, error) {
	var issueTypeModels []*model.IssueTypeModel
	if err := i.db.WithContext(ctx).
		Where("tenant_id = ? AND deleted_at IS NULL", tenantID).
		Order("hierarchy_level ASC").
		Find(&issueTypeModels).Error; err != nil {
		return nil, err
	}
	return i.mapper.ToEntities(issueTypeModels), nil
}

func (i *IssueTypeAdapter) SoftDeleteIssueType(ctx context.Context, tx *gorm.DB, issueTypeID int64, tenantID int64) error {
	db := i.getDB(tx)
	if err := db.WithContext(ctx).
		Model(&model.IssueTypeModel{}).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", issueTypeID, tenantID).
		Update("deleted_at", gorm.Expr("CURRENT_TIMESTAMP")).Error; err != nil {
		return err
	}
	return nil
}

func (i *IssueTypeAdapter) UpdateIssueType(ctx context.Context, tx *gorm.DB, issueType *entity.IssueTypeEntity) error {
	db := i.getDB(tx)
	issueTypeModel := i.mapper.ToModel(issueType)
	if err := db.WithContext(ctx).Save(issueTypeModel).Error; err != nil {
		return err
	}
	return nil
}

func (a *IssueTypeAdapter) getDB(tx *gorm.DB) *gorm.DB {
	if tx != nil {
		return tx
	}
	return a.db
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/adapter/IssueTypeAdapter.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.adapter;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import serp.project.pmcore.domain.issuetype.entity.IssueTypeEntity;
import serp.project.pmcore.domain.issuetype.port.IIssueTypePort;
import serp.project.pmcore.domain.issuetype.query.IssueTypeListCriteria;
import serp.project.pmcore.domain.shared.pagination.PageResult;
import serp.project.pmcore.infrastructure.store.mapper.IssueTypeMapper;
import serp.project.pmcore.infrastructure.store.model.IssueTypeModel;
import serp.project.pmcore.infrastructure.store.repository.IIssueTypeRepository;
import serp.project.pmcore.infrastructure.store.support.PageableUtils;

import java.util.List;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class IssueTypeAdapter implements IIssueTypePort {

    private final IIssueTypeRepository issueTypeRepository;
    private final IssueTypeMapper issueTypeMapper;

    @Override
    public IssueTypeEntity createIssueType(IssueTypeEntity issueType) {
        return issueTypeMapper.toEntity(issueTypeRepository.save(issueTypeMapper.toModel(issueType)));
    }

    @Override
    public Optional<IssueTypeEntity> getIssueTypeById(Long issueTypeId, Long tenantId) {
        return issueTypeRepository.findByIdAndTenantId(issueTypeId, tenantId)
                .map(issueTypeMapper::toEntity);
    }

    @Override
    public Optional<IssueTypeEntity> getIssueTypeByIdIncludingSystem(Long issueTypeId, Long tenantId) {
        return issueTypeRepository.findByIdAndTenantIdOrSystemTenant(issueTypeId, tenantId)
                .map(issueTypeMapper::toEntity);
    }

    @Override
    public Optional<IssueTypeEntity> getIssueTypeByTypeKey(Long tenantId, String typeKey) {
        return issueTypeRepository.findFirstByTenantIdAndTypeKeyOrderByIdAsc(tenantId, typeKey)
                .map(issueTypeMapper::toEntity);
    }

    @Override
    public Optional<IssueTypeEntity> getIssueTypeByTypeKeyIncludingSystem(Long tenantId, String typeKey) {
        return issueTypeRepository.findByTypeKeyAndTenantIdOrSystemTenant(typeKey, tenantId)
                .stream()
                .findFirst()
                .map(issueTypeMapper::toEntity);
    }

    @Override
    public List<IssueTypeEntity> listIssueTypes(Long tenantId) {
        return issueTypeMapper.toEntities(issueTypeRepository.findAllByTenantIdOrderByHierarchyLevelAsc(tenantId));
    }

    @Override
    public List<IssueTypeEntity> listIssueTypesIncludingSystem(Long tenantId) {
        return issueTypeMapper.toEntities(issueTypeRepository.findAllByTenantIdOrSystemTenant(tenantId));
    }

    @Override
    public List<IssueTypeEntity> getIssueTypesByIdsIncludingSystem(List<Long> issueTypeIds, Long tenantId) {
        if (issueTypeIds == null || issueTypeIds.isEmpty()) {
            return List.of();
        }
        return issueTypeMapper.toEntities(
                issueTypeRepository.findAllByIdInAndTenantIdOrSystemTenant(issueTypeIds, tenantId)
        );
    }

    @Override
    public PageResult<IssueTypeEntity> listIssueTypesIncludingSystem(Long tenantId, IssueTypeListCriteria criteria) {
        Pageable pageable = PageableUtils.of(criteria, resolveSort(criteria));
        Page<IssueTypeModel> result = issueTypeRepository.findAllVisibleWithFilters(
                tenantId,
                criteria.getSearch(),
                criteria.getHierarchyLevel(),
                criteria.getIsSystem(),
                pageable
        );
        return new PageResult<>(issueTypeMapper.toEntities(result.getContent()), result.getTotalElements());
    }

    @Override
    public void updateIssueType(IssueTypeEntity issueType) {
        issueTypeRepository.save(issueTypeMapper.toModel(issueType));
    }

    @Override
    public void deleteIssueType(Long issueTypeId, Long tenantId) {
        issueTypeRepository.deleteByIdAndTenantId(issueTypeId, tenantId);
    }

    @Override
    public boolean existsByTypeKey(Long tenantId, String typeKey) {
        return issueTypeRepository.existsByTenantIdAndTypeKey(tenantId, typeKey);
    }

    private Sort resolveSort(IssueTypeListCriteria criteria) {
        String sortBy = criteria.getSortBy().toLowerCase();
        Sort.Direction direction = PageableUtils.resolveDirection(criteria.getSortDirection());

        return switch (sortBy) {
            case "hierarchy_level" -> Sort.by(
                    new Sort.Order(direction, "hierarchyLevel"),
                    new Sort.Order(direction, "name"),
                    new Sort.Order(direction, "id")
            );
            case "name" -> Sort.by(
                    new Sort.Order(direction, "name"),
                    new Sort.Order(direction, "id")
            );
            case "type_key" -> Sort.by(
                    new Sort.Order(direction, "typeKey"),
                    new Sort.Order(direction, "id")
            );
            case "created_at" -> Sort.by(
                    new Sort.Order(direction, "createdAt"),
                    new Sort.Order(direction, "id")
            );
            default -> throw new IllegalArgumentException("sortBy must be one of hierarchy_level, name, type_key, created_at");
        };
    }
}
